src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -13,33 +13,49 @@ internal sealed class ModuleBuilderImpl : ModuleBuilder
     {
         private readonly Assembly _coreAssembly;
         private readonly string _name;
-        private readonly Dictionary<string, Type> _coreTypes = new();
+        private Type?[] _coreTypes;
         private readonly Dictionary<Assembly, AssemblyReferenceHandle> _assemblyRefStore = new();
         private readonly Dictionary<Type, TypeReferenceHandle> _typeRefStore = new();
         private readonly List<TypeBuilderImpl> _typeDefStore = new();
         private int _nextMethodDefRowId = 1;
         private int _nextFieldDefRowId = 1;
-
+        private static readonly Type[] s_coreTypes = { typeof(void), typeof(object), typeof(bool), typeof(char), typeof(sbyte), typeof(byte), typeof(short), typeof(ushort), typeof(int),
+                                                        typeof(uint), typeof(long), typeof(ulong), typeof(float), typeof(double), typeof(string), typeof(nint), typeof(nuint) };
         internal ModuleBuilderImpl(string name, Assembly coreAssembly)
         {
             _coreAssembly = coreAssembly;
+            // Use s_coreTypes directly for runtime reflection
+            _coreTypes = (_coreAssembly == typeof(object).Assembly) ? s_coreTypes : new Type[s_coreTypes.Length];
             _name = name;
         }
 
-        internal Type GetTypeFromCoreAssembly(string name)
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "Types are preserved via s_coreTypes")]
+        internal Type GetTypeFromCoreAssembly(CoreTypeId typeId)
+        {
+            int index = (int)typeId;
+
+            return _coreTypes[index] ?? (_coreTypes[index] = _coreAssembly.GetType(s_coreTypes[index].FullName!, throwOnError: true)!);
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "Types are preserved via s_coreTypes")]
```

Comment:
The justification doesn't make sense in this case. The `GetType` needs the type name passed in to be preserved, since we're accepting potentially any type as input here, the types listed in `s_coreTypes` are not going to fix this.
That said, the fact that we take the type as `System.Type` input means, that the type has to exist in the application (not trimmed away) and thus calling GetType on its full name should be fine. I know it is fine for normal trimming, and I "think" it's fine for AOT.

@MichalStrehovsky does AOT guarantee that `Type type = GetItSomewhere(); Type.GetType(type.FullName);` is compatible in all cases? That is if there is an instance of `System.Type` then that type can be looked up via `Type.GetType`.

Reply:
If I understand it correctly, this would not be reachable if we're operating on top of types that were subject to trimming. Per above, if `_coreAssembly == typeof(object).Assembly` is true, we would use a pre-populated array and not look up things by name. The lookup by name will only happen if we're emitting against a non-runtime type universe.

Reply:
> Do we need the GetType to work on all types, or is it only required (in this context) to work on the types listed in s_coreTypes? 

Yes, it is only need to work with types listed in s_coreTypes.

> If the latter, then this should work assuming there's a reference from this method to s_coreTypes. Currently there is not, so we would need to add that.

Ah, I see ...